Emit error when move-only type is used as a generic type

kavon opened the pull request:
resolves rdar://101651405

Needed-before-merge TODOs:
- [x] Prevent conformances to `Copyable` from appearing in swiftinterface files (see `Sema/missing-import-typealias.swift`)
- [x] Fix any other issues with other tests.

Deferrable TODOs:
- [x] Diagnose expressions that explicitly cast a move-only value to some existential as an error, not warning. i.e., `mo as! AnyHashable`. Alternatively, make sure they truly do fail at runtime.
- [ ] Improve diagnostics (specify the 2nd type it failed to match with in the message!)
- [ ] Prevent move-only types from appearing in tuples.
- [x] Make pack types (`repeat each T`) conform to `_Copyable` (see `Generics/tuple-conformances.swift`)
- [x] Get `_Copyable` out of the stdlib and make it some sort of builtin that's pre-populated in the ASTContext, like `Any`. We can't bootstrap the stdlib if that protocol is defined in some part of the stdlib, because every generic parameter is implicitly constrained to it and we have to be able to typecheck with `-parse-stdlib`. (see `Sema/stdlib_sugar_types.swift`)
- [ ] Determine whether changes to the diagnostics for unqualified member lookups are a significant concern and whether we can mitigate them. Example:
```
protocol MungedString {}

func format<T: MungedString>(_ t: T) {}
func takeGeneric<T>(_ t: T) {}

func caller() {
  // PREVIOUSLY error: type 'MungedString' has no member 'emptyString'
  // NOW error: reference to member 'emptyString' cannot be resolved without a contextual type
  format(.emptyString)

  // PREVIOUSLY error: cannot infer contextual base in reference to member 'somebody'
  // NOW error: type '_Copyable' has no member 'somebody'
  takeGeneric(.somebody)
}
```

kavon committed:
Initial ban of move-only types from being used generically

Since values of generic type are currently assumed to always
support copying, we need to prevent move-only types from
being substituted for generic type parameters.

This approach leans on a `_Copyable` marker protocol to which
all generic type parameters implicitly must conform.

A few other changes in this initial implementation:

- Now every concrete type that can conform to Copyable will do so. This fixes issues with conforming to a protocol that requires Copyable.
- Narrowly ban writing a concrete type `[T]` when `T` is move-only.

kavon committed:
functions, builtins, metatypes and some kinds of tuples to conform to `_Copyable`

kavon committed:
prevent move-only tuples from matching with `Any`

`Any`, which is effectively `any _Copyable`, is obviously copyable.

kavon committed:
move-only types can't witness associatedtype requirements

kavon committed:
all existentials are `_Copyable`

kavon commented:
@swift-ci please test

kavon committed:
prevent MO types from matching with existentials in the solver

`matchExistentialTypes` ends up getting called in most cases,
to ensure the conversion is legal.

kavon committed:
Ban explicit casts of move-only types unless its the identity cast

Since structs and enums only participate in subtyping via protocols,
move-only types that are structs and enums effectively are only
subtypes of themselves.

kavon commented:
@swift-ci please test

kavon committed:
[nfc] improve some comments

kavon committed:
slight tweak to how implicit conformance of nominals is generated

Also, only generate those conformances when the move-only feature is on,
for now.

kavon commented:
@swift-ci pls smoke test macOS platform

kavon commented:
@swift-ci please smoke test macOS platform

kavon committed:
prevent `MO as AnyObject` and friends from typechecking

where `MO` is a move-only type.

turns out it was being allowed because the cast is represented as a
disjunction in the constraint solver:

- `MO conv AnyObject`
- `MO bridge conv AnyObject`

I caught the first one but missed the second, which was unconditionally
being allowed.

kavon commented:
@swift-ci please test and merge

xedin commented on include/swift/Sema/CSFix.h:
```diff
@@ -2027,6 +2030,29 @@ class NotCompileTimeConst final : public ContextualMismatch {
+
+  bool diagnose(const Solution &solution, bool asNote = false) const override;
+
+  bool diagnoseForAmbiguity(CommonFixesArray commonFixes) const override {
```
Should be very careful with this because having the same location doesn’t mean that all fixes point to the same type.

kavon replied:
After looking at some of the other `ContextualMismatch::diagnoseForAmbiguity`, I think I know what you mean: the types within the fixes across the various solution-fix pairs might be different, so I should verify they're all the same first, otherwise return `false` to indicate that I can't diagnose this ambiguous set of solutions so that at least the "ambiguous solutions" error is emitted.

xedin replied:
Yeah, some of the fixes are like this, contextual is a good example.

kavon replied:
Thanks. Just fixed it.

kavon commented:
@swift-ci Please Test Source Compatibility

kavon commented:
@swift-ci please test

kavon committed:
make `MustBeCopyable::diagnoseForAmbiguity` robust

Previous implementation blindly took the first solution
and tried to diagnose, but that's not a safe assumption.

It's possible that among the solution-fix pairs, one
of `noncopyableTy`'s within the fix differs from the
others. Things probably can go wrong because the solution
doesn't correspond to that type.

Also, in that case we'd be emitting a diagnostic that
may not make any sense. In such cases, now we decline
to emit a diagnostic.

Thanks Pavel for catching this.

kavon commented:
@swift-ci please smoke test and merge

kavon committed:
update existing tests after banning move-only types in generics

A number of our existing tests put move-only types in optionals and
used them as arguments to generic type parameters. It turns out that
many of these appearances in the tests were not crucial for the test's
functionality itself.

The little coverage for force-unwrapping an Optional<moveOnly>
that was lost will be regained once we make these types sound.

I also tried to tidy up some of the tests with consistent naming for
operations on values, i.e., `consumeVal` and `borrowVal` functions.

kavon commented:
@swift-ci please smoke test and merge
